hathor/event/websocket/protocol.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2022 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Optional
+
+from autobahn.twisted.websocket import WebSocketServerProtocol
+from autobahn.websocket import ConnectionRequest
+from pydantic import ValidationError
+from structlog import get_logger
+
+from hathor.event.websocket.request import StreamRequest
+
+if TYPE_CHECKING:
+    from hathor.event.websocket import EventWebsocketFactory
+
+logger = get_logger()
+
+
+class EventWebsocketProtocol(WebSocketServerProtocol):
+    """ Websocket protocol, basically forwards some events to the Websocket factory.
+    """
+
+    factory: EventWebsocketFactory
+    client_peer: Optional[str] = None
+    last_received_event_id: Optional[int] = None
+    available_window_size: int = 0
+
+    @property
+    def next_event_id(self) -> int:
+        """The ID of the next event the client expects."""
```

Comment:
If I got it right, the `last_received_event_id` refers to the last ACK sent by the client. It's different from the next event to be sent to the client. Let's say we have a window size of 100 and I'm starting at index 1. I might have already sent 1, 2, ..., 43 while the client had acknowledged only up to event 24. In this case, I should keep going and send event 44 (regardless of the last ack received).

Besides that, I'd suggest you to rename the method to `next_expected_event_id(...)`.

Lastly, I don't feel it's useful to have it as a `@property`. But I don't have strong arguments to support my feeling so I'm just sharing my intuition about it.

Reply:
> If I got it right, the `last_received_event_id` refers to the last ACK sent by the client.

Correct, but also every time the server sends an event to the client, `last_received_event_id` is updated to that event's ID. Then, when the client sends a new ACK, it replaces the value tracked by the server.

Maybe changing the name to `last_sent_event_id` would be clearer?

> (...) In this case, I should keep going and send event 44 (regardless of the last ack received).

That works correctly because of the described above.

> Besides that, I'd suggest you to rename the method to `next_expected_event_id(...)`.

Done!

> Lastly, I don't feel it's useful to have it as a `@property`. But I don't have strong arguments to support my feeling so I'm just sharing my intuition about it.

Ok, I have changed it to a method.

Reply:
Thanks!